### PR TITLE
feat(jenkins-backend): extend config to allow passing extra headers

### PR DESCRIPTION
**Pull request by tanguyantoine**
Signed-off-by: TANGUY Antoine  <tanguyantoine@users.noreply.github.com>

## Hey, I just made a Pull Request!

- extended `jenkins-backend` plugin configuration so that we can send extra headers to jenkins

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] ~~Screenshots attached (for UI changes)~~ not applicable
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
Oh just a curious question too - you didn't mention a rationale here; what is it that you need to pass in? Could this have been done with a proxy route that injects headers instead?

**Comment by tanguyantoine** [comment]
We have an instance protected and every request must contain proper headers. 
If this plugin let us send extra headers, we can avoid deploying a proxy

**Comment by freben** [comment]
Right. I just meant that Backstage already comes with a proxy builtin, and that's used by a lot of plugins. It's just a matter of adding a few lines to the config file and you're up and running. Your approach works, but is very "targeted"; it solves one very narrow feature for one single plugin, and if others have slightly different needs then this will grow over time.

File: plugins/jenkins-backend/README.md [hunk 1]
```diff
@@ -89,6 +89,9 @@ jenkins:
   baseUrl: https://jenkins.example.com
   username: backstage-bot
   apiKey: 123456789abcdef0123456789abcedf012
+  # optionally add extra headers
+  # headers:
```
Comment (freben) [inline plugins/jenkins-backend/README.md:93]: remember to change this one to `extraRequestHeaders`

File: plugins/jenkins-backend/src/service/jenkinsInfoProvider.test.ts [hunk 2]
```diff
@@ -99,12 +106,14 @@ describe('JenkinsConfig', () => {
         baseUrl: 'https://jenkins.example.com',
         username: 'backstage - bot',
         apiKey: '123456789abcdef0123456789abcedf012',
+        headers: {},
```
Comment (freben) [inline plugins/jenkins-backend/src/service/jenkinsInfoProvider.test.ts:109]: now you don't need these empty ones

File: plugins/jenkins-backend/src/service/jenkinsInfoProvider.ts [hunk 3]
```diff
@@ -74,12 +78,13 @@ export class JenkinsConfig {
...
 
     // load all named instance config
-    const namedInstanceConfig =
+    const namedInstanceConfig: JenkinsInstanceConfig[] =
       jenkinsConfig.getOptionalConfigArray('instances')?.map(c => ({
         name: c.getString('name'),
         baseUrl: c.getString('baseUrl'),
         username: c.getString('username'),
         apiKey: c.getString('apiKey'),
+        headers: c.getOptional('headers') || {},
```
Comment (freben) [inline plugins/jenkins-backend/src/service/jenkinsInfoProvider.ts:87]: and you no longer need the `|| {}` either

File: plugins/jenkins-backend/src/service/jenkinsInfoProvider.ts [hunk 4]
```diff
@@ -93,6 +98,10 @@ export class JenkinsConfig {
     const username = jenkinsConfig.getOptionalString('username');
     const apiKey = jenkinsConfig.getOptionalString('apiKey');
     const crumbIssuer = jenkinsConfig.getOptionalBoolean('crumbIssuer');
+    const extraRequestHeaders =
+      jenkinsConfig.getOptional<JenkinsInstanceConfig['extraRequestHeaders']>(
+        'extraRequestHeaders',
+      ) || {};
```
Comment (freben) [inline plugins/jenkins-backend/src/service/jenkinsInfoProvider.ts:104]: nor here

**Review by freben — COMMENTED** [review]
aight, just the final tweaks now

**Comment by tanguyantoine** [comment]
Updated

**Review by freben — APPROVED** [review]
Alright!

**Pull request merged**